### DISCUSSION TITLE
Feedback on sqlalchemy2.0 + MappedAsDataclass

### DISCUSSION THREAD

**Discussion by arogozhnikov** [sqlalchemy/sqlalchemy#9394] [Ideas]
Hi, I’m testing how my codebase could work with sqlalchemy2.0 (currently sqlalchemy1.4 + sqlmodel) 

I’ve translated most of the parts, here is my feedback so far:

- typing + sqlalchemy = 🎉🎉🎉
- compared to sqlmodel, can do inheritance of tables, and less magic with not-so-common-fields
- minus one dependency just feels good

Downgrades:
- dataclass typing works with VS code, but not in pycharm. That’s completely pycharm’s fault, but that’s how it is
- sqlmodel’s `sess.exec(select(Entity)).all()` already typed as `list[Entity]`, no need for `.scalars`, no need to convert Sequence to list
- sqlmodel’s `sess.exec(select(Entity1, Entity2)).all()` already typed as `list[tuple[Entity1, Entity2]]`. I don't see benefits for typing result as `Row[_Tuple]`, not `_Tuple`. I find `.t/.tuple()` annoying, and ended up writing custom function `get_all_tuples(session, statement)` which returns `list[_Tuple]`
- Next group of complains is mostly about link properties (`parent: Mapped[Parent] = relationship()`)
    - `__repr__` is strange… it tries to pull those fields and thus fails for detached objects. It also does this in circle (parent -> children -> parent -> children) and produces unreadable result. Better not having links in repr
    - not that I need pydantic, but it is nice to have .dict() method that collects all fields. Asdict additionally pulls link properties, which I don’t think should be there. In general, I don’t think link properties should be treated as normal dataclass properties 
    - sqlmodel’s ability to accept either `parent_id` or `parent` in constructor is actually convenient when creating a number of entities within list comprehension `[Child(name=name, parent=parent) for name in names]`.  I think some method like `[Child(name=name).link(parent=parent) for name in names]` would be ok and can maintain separation between fields and link properties.

**Comment by CaselIT** [comment]
Hi,

thanks for the feedback, they are appreciated. some comments below.

> dataclass typing works with VS code, but not in pycharm. That’s completely pycharm’s fault, but that’s how it is

pycharm seems to have its how type checker that a bit of an hit and miss. I'm not sure if there is anything sqlalchemy can do to improve it. hopefully it will get better once pycharm has time to accommodate the sqlalcehmy update. 

> - sqlmodel’s sess.exec(select(Entity)).all() already typed as list[Entity], no need for .scalars, no need to convert Sequence to list

I personally like the consistency of the execute. You always get rows. In a generic function you otherwise don't know if you are getting back rows or scalars. Explicit is better than implicit, or so says python. You can use `session.scalars()` were you need a list of scalars.

Regarding sequence to list, I don't understand why you need to convert to list. Are you appending things to the values returned from an result? What's missing in the sequence interface that requires a list?

> sqlmodel’s sess.exec(select(Entity1, Entity2)).all() already typed as list[tuple[Entity1, Entity2]]. I don't see benefits for typing result as Row[_Tuple], not _Tuple. I find .t/.tuple() annoying, and ended up writing custom function get_all_tuples(session, statement) which returns list[_Tuple]

that's the incorrect type though. You get back row objects that are more like named tuples than tuples. You also can transform Row to RowMapping to have a mapping like. You can't do any of that if the return type were tuples.

Why do you prefer `sequence[tuple]`?

> Next group of complains is mostly about link properties (parent: Mapped[Parent] = relationship())

the following seems more feedbacks on dataclasses that sqlalchemy. sqlalchemy tries to do use them without any change

> __repr__ is strange… it tries to pull those fields and thus fails for detached objects. It also does this in circle (parent -> children -> parent -> children) and produces unreadable result. Better not having links in repr

this the dataclass `__repr__`. You can use `relationship(repr=False)` to omit it from the repr.
it will also cause lazy loads, since it does not care about what's in the class `__dict__`, it will always access all attributes that have repr=True

> not that I need pydantic, but it is nice to have .dict() method that collects all fields. Asdict additionally pulls link properties, which I don’t think should be there. In general, I don’t think link properties should be treated as normal dataclass properties

I'm not sure if it's something that sqlalchemy has control of. dataclass have one kind of property, so you either get typing or you don't. you may use dataclass callable (see below) to customize what fed to the dataclass function

> sqlmodel’s ability to accept either parent_id or parent in constructor is actually convenient when creating a number of entities within list comprehension [Child(name=name, parent=parent) for name in names].  I think some method like [Child(name=name).link(parent=parent) for name in names] would be ok and can maintain separation between fields and link properties.

that's also how the dataclass init works. An option is to have `init=False` in the `parent_id` or `default=None`. Currently there isn't a safe default for relationships, since `None` means no-relation, and it has priority over the `parent_id` fk, so even if parent_id is set by the constructor, it's nulled out by the relationship being set to `None`.
If you omit the dataclass init, using the class kwargs `init=False` you get the default sqlalchemy one, that works similarly to the sqlmodel one, but it's typed as `__init__(self, **kwargs)` so you do lose typing in the constructor

This was raised in other issues/discussion, and a solution may be to have a special token `Missing` (or something along these lines) that's understood by sqlalchemy to mean "not set". But that complicates typing (since you need to type `Mapped[list[x] | Literal[Missing]]`. cc @zzzeek 

Note that since 2.0.4 you can pass your own dataclass callable, to use something like pydantic, or to customize some of the class attributes before feeding it to the default dataclass callable. https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#integrating-with-alternate-dataclass-providers-such-as-pydantic

Again thanks for testing and the feedback.

**Reply by arogozhnikov** [reply]
Hi @CaselIT, thanks for detailed response!

> What's missing in the sequence interface that requires a list?

brevity of typing, e.g. I commonly use list[Entity] for inputs in function. Passing Sequence[Entity] raises warnings, so as a fastest solution I wrap result in list.

> Why do you prefer sequence[tuple]?

In types I don't see names of fields, here is what I do:
```python
child_with_parent = get_all_tuples(session, select(Child, Parent).join(Child.parent))
parent_id2count = Counter(p.parent_id for _, p in child_with_parent)
child_id2parent_id = {c.child_id:p.parent_id for c, p in child_with_parent} 
... more operations
```
in this case all fields are completely typed, and if I did something wrong - I'll see warning.
Also I can't see a good way to use `.t` within comprehension.

> dataclass have one kind of property

I believe there are ways to make them not properties. That's my point - link properties should be like getter/setter, and not 'normal' fields of dataclasses. 'Normal' fields should be similar to contents of row in the database.

Dataclasses documentation defines pseudo-fields (InitVar, ClassVar) -  I think ClassVar is a good solution (typed, accessible as field, not used in asdict and repr). Not available in ctor though.

**Reply by CaselIT** [reply]
> > Why do you prefer sequence[tuple]?
> 
> In types I don't see names of fields, here is what I do:
> 
> ```python
> child_with_parent = get_all_tuples(session, select(Child, Parent).join(Child.parent))
> parent_id2count = Counter(p.parent_id for _, p in child_with_parent)
> child_id2parent_id = {c.child_id:p.parent_id for c, p in child_with_parent} 
> ... more operations
> ```
> 
> in this case all fields are completely typed, and if I did something wrong - I'll see warning. Also I can't see a good way to use `.t` within comprehension.

Yes, there is currently no way of having an object that's not a tuple behave like a tuple that has a different type on each index, so the pattern ` for _, p` only works for tuples. That's what `.t` is for. This is one of a python typing many limitation. IIRC there is an issue in the python/typing repo so there may be a pep to support it one that. see mike post on it here https://github.com/sqlalchemy/sqlalchemy/discussions/9394#discussioncomment-5166612
In the meantime if `child_with_parent ` is a plain Result you can use `child_with_parent.t.all()` instead of `child_with_parent.all()` to get back a sequence of tuples that you can act upon in a typed way. I don't think it's a particular issue having to type the `.t` when you need destructoring to work.

> 
> > dataclass have one kind of property
> 
> I believe there are ways to make them not properties. That's my point - link properties should be like getter/setter, and not 'normal' fields of dataclasses. 'Normal' fields should be similar to contents of row in the database.

well, sure, but then they are not properties, hence my "dataclass have one kind of property". You either use it or not.
Personally I disagree that they should not be property here, since they are more like a property that not.

The main problem here is the limitation of dataclasses. I think pydantic solve many of them by looking at what's in `__dict__` and only acting on it. dataclass has no concept of "partially populated", so it will always repr or serialize every field.

> 
> Dataclasses documentation defines pseudo-fields (InitVar, ClassVar) - I think ClassVar is a good solution (typed, accessible as field, not used in asdict and repr). Not available in ctor though.

But they are not class var. They are present in the instance with a different value than when accessed in the class. They also behave like normal property in every other way.

Also I don't think sqlalchemy complains if you do `ClassVar[Mapped[...]]`, but I haven't tried it. That is users would need to opt into using classvars.

**Reply by arogozhnikov** [reply]
Somehow missed I can use .t on result, not row. Thanks!

> But they are not class var.

Sure. But ClassVar's behavior in runtime and typing is more predictable and convenient than any other alternative I have right now.

>  Personally I disagree that they should not be property here, since they are more like a property that not.

I've just checked how sqlmodel deals with it. Link properties are not a part of pydantic model - they are not registered in `__fields__`. 

Given that link is lazy-loaded, that doesn't feel like a field of dataclass and has issues I've reported.

What are benefits of having link as a property?

**Reply by CaselIT** [reply]
> I've just checked how sqlmodel deals with it. Link properties are not a part of pydantic model - they are not registered in __fields__.

interesting, does pydantic allows whatever in it's init and it just sets it in the dict? How does setting a relationship in the constructor work?

> Given that link is lazy-loaded, that doesn't feel like a field of dataclass and has issues I've reported.

well they can be eagerly loaded

> What are benefits of having link as a property?

that you can set them in the init

> Also I don't think sqlalchemy complains if you do ClassVar[Mapped[...]], but I haven't tried it. That is users would need to opt into using classvars.

I've just tried and it does not work with a bare `relationship()`, but if you pass the type to the relationship everything works ok. There are no test but if we get requests on this pattern we may better accommodate it
```py
import sqlalchemy as sa
from sqlalchemy import orm
from typing import ClassVar

class B(orm.DeclarativeBase, orm.MappedAsDataclass):
    pass

class X(B):
    __tablename__ = "x"
    id: orm.Mapped[int] = orm.mapped_column(primary_key=True)
    yid: orm.Mapped[int] = orm.mapped_column(sa.ForeignKey("y.id"))

class Y(B):
    __tablename__ = "y"
    id: orm.Mapped[int] = orm.mapped_column(primary_key=True)
    xs: ClassVar[orm.Mapped[list[X]]] = orm.relationship(X)


print(sa.select(Y))
print(sa.select(Y).where(Y.xs.any(X.id > 1)))
print(Y(123))
```

note that this is something the user needs to opt in, it's not something that can be done on sqlalchemy side (other than maybe improving the type parsing logic to support that pattern without the need of providing a value to relationship)

**Reply by arogozhnikov** [reply]
> but if you pass the type to the relationship everything works ok.

Yup, that's what I currently use.

> that you can set them in the init

Pros for being field:
+ can be set in init

Cons:
- mess up repr (fixable)
- mess up asdict (not fixable)
- in case of loops in properties (like parent/children), `asdict` fails with infinite recursion. Convincing argument that link is not really a field
- if used in constructor, there is currently a conflict with setting foreign keys. Ignoring previous points, I couldn't opt into behavior where I always set link, and not foreign_id

> interesting, does pydantic allows whatever in it's init and it just sets it in the dict? How does setting a relationship in the constructor work?

Yes, but with pydantic-style prefiltering. De-facto init accepts **kwargs.
SQLmodel fools typers to believe it is a pydantic model, so they suggest standard pydantic constructor with links included.
Relationship is just set in `__dict__` as one more attribute, and if it is not passed in kwargs - it is not set.

**Comment by zzzeek** [comment]
`Row` has all kinds of cool features, not the least of which is that it's a named tuple also, which is also not a behavior we can drop.     You have to square the use case of `sess.exec(select(Entity1, Entity2)).all()` with `sess.exec(select(Entity1.id, Entity1.name)).all()`, that is, columns and not entities.    `Row` provides one consistent interface to all of them.       The vast majority of our users are looking for functionality that's been there for a decade.   The `.t` thing can be fixed if and when Python finally fixes [pep-646](https://peps.python.org/pep-0646/) to be useful.   That's where the commentary should go on this IMO, major failing in current Python typing.

**Reply by zzzeek** [reply]
Here's what we need from pep-646 - any ability whatsoever to refer to unpacking the actual elements:

```py
from typing import Generic
from typing import TYPE_CHECKING

from typing_extensions import TypeVarTuple
from typing_extensions import Unpack


R = TypeVarTuple("R")

class Row(Generic[Unpack[R]]):
    ...

    # does not work, each element gets the full Tuple
    # def __iter__(self) -> Iterator[Tuple[Unpack[R]]]:
    #     ...


if TYPE_CHECKING:
    r: Row[int, str] = Row()

    # does not work, no __iter__
    c, d = r

```

if they can fix that, then we have everything.  we are waiting on that.

**Reply by arogozhnikov** [reply]
I see. Yes, that's a bit weird

**Reply by zzzeek** [reply]
yeah it's like they spent a year on a pep that is mostly useless

**Reply by CaselIT** [reply]
do you know if there is an issue to point out that that pep misses the mark?

**Reply by zzzeek** [reply]
I've pointed it out on the typing list and in typing discussions, yes, like here: https://mail.python.org/archives/list/typing-sig@python.org/thread/TWHKYQMDKBXXXIRVWRQZCCAZ52E3YPFI/#DFTXPC2Z5XPCG7LEDJANOCUXVIAE7RGA .   there's other more murky peps being worked on...somewhere.
